File: FruitWholesale.Mobile/lib/features/ledgers/ledger_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'ledger_models.dart';

class LedgerService {
  final ApiClient _api;

  LedgerService(this._api);

  Future<PaginatedList<LedgerEntry>> getShopLedger(int shopId, {int pageNumber = 1}) async {
    final json = await _api.get('/ledger/shop/$shopId', query: {'pageNumber': pageNumber, 'pageSize': 20});
    return PaginatedList.fromJson(json as Map<String, dynamic>, LedgerEntry.fromJson);
  }

  Future<PaginatedList<LedgerEntry>> getSupplierLedger(int supplierId, {int pageNumber = 1}) async {
    final json = await _api.get('/ledger/supplier/$supplierId', query: {'pageNumber': pageNumber, 'pageSize': 20});
    return PaginatedList.fromJson(json as Map<String, dynamic>, LedgerEntry.fromJson);
  }

  Future<PaginatedList<CashLedgerEntry>> getCashLedger({int pageNumber = 1}) async {
    final json = await _api.get('/ledger/cash', query: {'pageNumber': pageNumber, 'pageSize': 20});
    return PaginatedList.fromJson(json as Map<String, dynamic>, CashLedgerEntry.fromJson);
  }
}

File: FruitWholesale.Client/src/app/features/ledgers/ledger.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { CashLedgerEntry, ShopLedgerEntry, SupplierLedgerEntry } from '../../core/models/ledger.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class LedgerService {
  private readonly baseUrl = `${environment.apiUrl}/ledger`;

  constructor(private readonly http: HttpClient) {}

  getShopLedger(shopId: number, request: PaginationRequest, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<ShopLedgerEntry>>(`${this.baseUrl}/shop/${shopId}`, {
      params: toHttpParams({ ...request, fromDate, toDate })
    });
  }

  getSupplierLedger(supplierId: number, request: PaginationRequest, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<SupplierLedgerEntry>>(`${this.baseUrl}/supplier/${supplierId}`, {
      params: toHttpParams({ ...request, fromDate, toDate })
    });
  }

  getCashLedger(request: PaginationRequest, fromDate?: string | null, toDate?: string | null, transactionType?: string | null) {
    return this.http.get<PaginatedList<CashLedgerEntry>>(`${this.baseUrl}/cash`, {
      params: toHttpParams({ ...request, fromDate, toDate, transactionType })
    });
  }
}
